ci-operator/step-registry/baremetal/lab/agent/day2/baremetal-lab-agent-day2-commands.sh

```diff
@@ -0,0 +1,181 @@
+  oc get secret -n openshift-config pull-secret -o jsonpath='{.data.\.dockerconfigjson}' | base64 -d > "${day2_pull_secret}"
+fi
+
+DAY2_INSTALL_DIR="${DAY2_INSTALL_DIR:-/tmp/installer_day2}"
+mkdir -p "${DAY2_INSTALL_DIR}"
+cp "${SHARED_DIR}/nodes-config.yaml" "${DAY2_INSTALL_DIR}/"
+cp "${SHARED_DIR}/nodes-config.yaml" "${ARTIFACT_DIR}/"
+
+echo "Create node.iso for day2 worker nodes..."
+/tmp/oc adm node-image create --dir="${DAY2_INSTALL_DIR}" -a "${day2_pull_secret}" --insecure=true
+
+CLUSTER_NAME=$(<"${SHARED_DIR}/cluster_name")
+
+gnu_arch=$(echo "$architecture" | sed 's/arm64/aarch64/;s/amd64/x86_64/;')
+case "${BOOT_MODE}" in
+"iso")
+  ### Copy the image to the auxiliary host
+  echo -e "\nCopying the day2 node ISO image into the bastion host..."
+  scp "${SSHOPTS[@]}" "${DAY2_INSTALL_DIR}/node.iso" "root@${AUX_HOST}:/opt/html/${CLUSTER_NAME}.node.iso"
+  echo -e "\nMounting the ISO image in the hosts via virtual media and powering on the hosts..."
+  # shellcheck disable=SC2154
+  for bmhost in $(yq e -o=j -I=0 '.[] | select(.name|test("-a-"))' "${SHARED_DIR}/hosts.yaml"); do
+    # shellcheck disable=SC1090
+    . <(echo "$bmhost" | yq e 'to_entries | .[] | (.key + "=\"" + .value + "\"")')
+    if [ "${transfer_protocol_type}" == "cifs" ]; then
+      IP_ADDRESS="$(dig +short "${AUX_HOST}")"
+      iso_path="${IP_ADDRESS}/isos/${CLUSTER_NAME}.node.iso"
+    else
+      # Assuming HTTP or HTTPS
+      iso_path="${transfer_protocol_type:-http}://${AUX_HOST}/${CLUSTER_NAME}.node.iso"
+    fi
+    mount_virtual_media "${host}" "${iso_path}"
+  done
+
+  wait
+  if [ -f /tmp/virtual_media_mount_failed ]; then
+    echo "Failed to mount the ISO image in one or more hosts"
+    exit 1
+  fi
+;;
+"pxe")
+  ### Create pxe files
+  echo -e "\nCreating PXE files..."
+  oinst agent create pxe-files
+  ### Copy the image to the auxiliary host
+  echo -e "\nCopying the PXE files into the bastion host..."
+  scp "${SSHOPTS[@]}" "${DAY2_INSTALL_DIR}"/boot-artifacts/agent.*-vmlinuz* \
+    "root@${AUX_HOST}:/opt/dnsmasq/tftpboot/${CLUSTER_NAME}/vmlinuz_${gnu_arch}"
+  scp "${SSHOPTS[@]}" "${DAY2_INSTALL_DIR}"/boot-artifacts/agent.*-initrd* \
+    "root@${AUX_HOST}:/opt/dnsmasq/tftpboot/${CLUSTER_NAME}/initramfs_${gnu_arch}.img"
+  scp "${SSHOPTS[@]}" "${DAY2_INSTALL_DIR}"/boot-artifacts/agent.*-rootfs* \
+    "root@${AUX_HOST}:/opt/html/${CLUSTER_NAME}/rootfs-${gnu_arch}.img"
+;;
+*)
+  echo "Unknown install mode: ${BOOT_MODE}"
+  exit 1
+esac
+
+# shellcheck disable=SC2154
+for bmhost in $(yq e -o=j -I=0 '.[] | select(.name|test("-a-"))' "${SHARED_DIR}/hosts.yaml"); do
+  # shellcheck disable=SC1090
+  . <(echo "$bmhost" | yq e 'to_entries | .[] | (.key + "=\"" + .value + "\"")')
+  echo "Power on #${host} (${name})..."
+  timeout -s 9 10m ssh "${SSHOPTS[@]}" "root@${AUX_HOST}" prepare_host_for_boot "${host}" "${BOOT_MODE}"
+done
+
+echo "Monitoring day2 workers and pending csrs..."
+
+monitor_workers &
+approve_csrs
+
```

Comment:
What's the purpose of first and second?

Reply:
The first and second texts will be logged in the monitoring logs for CSRs. For better readability in the condition, we can use the terms 'First CSR' and 'Second CSR.' Here's a sample of the logs:

```
time=2024-08-09T06:46:07Z level=info msg=Node 192.168.12.5: First CSR Pending approval
time=2024-08-09T06:46:07Z level=info msg=Node 192.168.12.5: CSR csr-vdgh7 with signerName kubernetes.io/kube-apiserver-client-kubelet and username system:serviceaccount:openshift-machine-config-operator:node-bootstrapper is Pending and awaiting approval
time=2024-08-09T06:48:55Z level=info msg=Node 192.168.12.5: Second CSR Pending approval
time=2024-08-09T06:48:55Z level=info msg=Node 192.168.12.5: CSR csr-rwztn with signerName kubernetes.io/kubelet-serving and username system:node:node3 is Pending and awaiting approval
```

Reply:
Why not just using
```
monitor_workers & # The current implementation
approve_csr & # This one may be kept foreground
wait # Wait for all the sub-processes to quit, not only approve_csr.
```

Reply:
This feature requires validating the monitoring command to provide information on pending CSRs. Although I know the current implementation will work but it does not validate the feature requirements. Since automatic installations might not always be closely monitored for expected information in the logs, I recommend reviewing the logs to ensure the necessary information is present before approving the pending CSRs. For more details, please refer to the [epic](https://issues.redhat.com/browse/AGENT-896).

Reply:
Ack, thanks for clarifying. At this point, we should not use approve_csrs at all.... Rather, we'd parse the output of the node-monitor command as in

```bash
/tmp/oc adm node-image monitor --ip-addresses "${DAY2_IPS}" -a "${day2_pull_secret}" --insecure=true 2>&1 | \
  tee output.txt | while IFS= read -r line; do
  if [[ "$line" = *"with signerName kubernetes.io/kube-apiserver-client-kubelet and username system:serviceaccount:openshift-machine-config-operator:node-bootstrapper is Pending and awaiting approval"* ]]; then
    node_ip=$(echo "$line" | sed 's/^.*Node \(.*\): CSR.*$/\1/')
    csr=$(echo "$line" | sed 's/^.*CSR \([^ ]*\).*$/\1/')
    echo "Approving CSR $csr for node $node_ip"
    oc adm certificate approve "$csr"
  fi
  if [[ "$line" = *"with signerName kubernetes.io/kubelet-serving and username "*" is Pending and awaiting approval"* ]]; then
    node_ip=$(echo "$line" | sed 's/^.*Node \(.*\): CSR.*$/\1/')
    csr=$(echo "$line" | sed 's/^.*CSR \([^ ]*\).*$/\1/')
    echo "Approving CSR $csr for node $node_ip"
    oc adm certificate approve "$csr"
  fi
done
EXIT_STATUS="${PIPESTATUS[0]}"
echo "Exiting with status $EXIT_STATUS"
exit $EXIT_STATUS
```

Reply:
Thanks for the suggestion. Yes, it meets the expectations and does not require the approve_csrs function. Also, it should be fine to use the first condition with `*"kube-apiserver-client-kubelet"*` and the second with `*"kubelet-serving"*` to approve the CSRs.